### PR TITLE
feat: prometheus row inserter

### PR DESCRIPTION
**Pull request by fengjiachun**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Prometheus remote write with row-based protocol

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/servers/src/row_writer.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            rows: if num_rows > 0 {
+                Vec::with_capacity(num_rows)
+            } else {
+                Vec::new()
+            },
+            column_indexes: if num_columns > 0 {
+                HashMap::with_capacity(num_columns)
+            } else {
+                HashMap::new()
+            },
```
Comment (waynexia) [inline src/servers/src/row_writer.rs:53]: You can simplify this to `with_capacity(num_x)` unconditionally -- it's equal to `new()` when capacity is 0 https://doc.rust-lang.org/src/alloc/raw_vec.rs.html#120-123

File: src/servers/src/row_writer.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        table_name: &'a str,
+        num_columns: usize,
+        num_rows: usize,
+    ) -> &mut TableData<'a> {
+        self.table_data_map
+            .entry(table_name)
+            .or_insert_with(|| TableData::new(num_columns, num_rows))
+    }
+
+    pub fn into_row_insert_requests(self) -> (RowInsertRequests, usize) {
```
Comment (waynexia) [inline src/servers/src/row_writer.rs:104]:
```suggestion
    /// Returns the request and number of rows in it.
    pub fn into_row_insert_requests(self) -> (RowInsertRequests, usize) {
```

File: src/servers/src/influxdb.rs
Comment (waynexia) [inline src/servers/src/influxdb.rs:1]: 👍

**Review by waynexia — APPROVED** [review]

File: src/servers/src/prom_store.rs [hunk 3]
```diff
@@ -300,6 +301,59 @@ fn recordbatch_to_timeseries(table: &str, recordbatch: RecordBatch) -> Result<Ve
...
+
+        for Sample { value, timestamp } in &series.samples {
+            let mut one_row = table_data.alloc_one_row();
+
+            // labels
+            let kvs = series
+                .labels
+                .iter()
+                .skip_while(|label| label.name == METRIC_NAME_LABEL)
+                .map(|label| (label.name.as_str(), label.value.as_str()));
```
Comment (evenyag) [inline src/servers/src/prom_store.rs:338]: issue: we should use filter_map. [skip_while()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip_while) always yield the element once the predicate returns false
```suggestion
                .filter_map(|label| if label.name == METRIC_NAME_LABEL { None } else { Some((label.name.as_str(), label.value.as_str()) });
```

The following code outputs `[1, -1, 0, 1]`
```rust
fn main() {
let a = [1, -1i32, 0, 1];
let b: Vec<_> = a.iter().skip_while(|x| x.is_negative()).collect();
println!("{b:?}");
}
```

File: src/servers/src/row_writer.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    one_row: &mut Vec<Value>,
+) -> Result<()> {
+    let TableData {
+        schema,
+        column_indexes,
+        ..
+    } = table_data;
+
+    let ts = match ts {
+        Some(timestamp) => writer::to_ms_ts(precision, timestamp),
```
Comment (evenyag) [inline src/servers/src/row_writer.rs:221]: problem(non-blocking): We might lose precision if users use influxdb protocol to write to us (Timestamp in line-protocol is in nanosecond precision).

File: src/servers/src/row_writer.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    pub fn into_row_insert_requests(self) -> (RowInsertRequests, usize) {
+        let mut total_rows = 0;
+        let inserts = self
+            .table_data_map
+            .into_iter()
+            .map(|(table_name, table_data)| {
+                total_rows += table_data.num_rows();
+                let num_columns = table_data.num_columns();
+                let (schema, mut rows) = table_data.into_schema_and_rows();
+                for row in rows.iter_mut() {
```
Comment (evenyag) [inline src/servers/src/row_writer.rs:113]:
```suggestion
                for row in &mut rows {
```

File: src/servers/src/row_writer.rs (hunk 4 above)
Comment (fengjiachun) [reply src/servers/src/row_writer.rs:221]: 是的，你说的对。我遵循了列插入的处理方式

Comment (fengjiachun) [reply src/servers/src/row_writer.rs:221]: Yes, you are right. However, let's keep the same processing as [column-based protocol](https://github.com/GreptimeTeam/greptimedb/blob/d335bfc7bfb494215e90b430d1b324e8c2ce345c/src/common/grpc/src/writer.rs#L66) for now. Perhaps we can consider making a unified modification to nanoseconds.

**Review by evenyag — APPROVED** [review]

**Pull request merged**